synapse/handlers/room.py

```diff
@@ -147,6 +147,9 @@ async def upgrade_room(
 
         # Check if this room is already being upgraded by another person
         for key in self._upgrade_response_cache.pending_result_cache:
+            # The keys of pending_result_cache just need to be hashable, but in
+            # this case we know that it is a tuple.
+            assert isinstance(key, tuple)
```

Comment:
I'd be somewhat tempted to make `ResponseCache` generic over the key type. Though I guess it isn't a big deal if we only do this in one place.

Reply:
I guess it is safe to assume that a particular cache always has the same key types, so that would probably make more sense! I'll take a look at how hard that is.